src/Bicep.Core/Registry/AzureContainerRegistryManager.cs

```diff
@@ -103,12 +105,45 @@ public async Task PushArtifactAsync(RootConfiguration configuration, OciArtifact
             var manifestUploadResult = await blobClient.UploadManifestAsync(manifestStream, new UploadManifestOptions(tag: moduleReference.Tag));
         }
 
+        public bool CheckArtifactExists(RootConfiguration configuration, OciArtifactModuleReference moduleReference)
+        {
+            var client = this.CreateRegistryClient(configuration, moduleReference);
+
+            try
+            {
+                Trace.WriteLine($"Authenticated attempt to check if artifact exists for module {moduleReference.FullyQualifiedReference}.");
+                // Try authenticated client first.
+                // Check if there is an artifact with the expected tag
+                var artifact = client.GetArtifact(moduleReference.Repository, moduleReference.Tag);
+                var tagInfo = artifact.GetTagProperties(moduleReference.Tag);
+                return artifact.GetAllTagProperties().Any(a => a.Name == moduleReference.Tag);
```

Comment:
>GetAllTagProperties [](http://example.com/codeflow?start=32&length=19)

The challenge with this approach is that you can verify that the tag exists, but it's not guaranteed that the tag actually points to a Bicep module.

I think `CheckModuleExists()` in `OciModuleRegistry` can just call `PullArtifactAsync` and validate that it got a manifest that belongs to a Bicep module. (It shouldn't touch the local module cache, though.)

If you get an exception with 404, then the module doesn't exist. If you get a manifest that isn't a Bicep module, then the user may want to know about that as well.

This approach would also remove the need to create separate registry clients as well. The existing blob client would be sufficient.

Reply:
I didn't consider that possible false positive. Great call, I'll do that!